msprime/demography.py

```diff
@@ -2431,6 +2459,301 @@ def _american_admixture_model():
+            ax.figure.savefig("island_model.pdf")
+
+        :return: A :class:`demes.Graph` object corresponding to the demography.
+        :rtype: demes.Graph
+        """
+        resolved = self.validate()
+        b = demes.Builder()
+        for pop in resolved.populations:
+            initial_epoch = dict(
+                end_size=pop.initial_size, growth_rate=pop.growth_rate, end_time=0
+            )
+            b.add_deme(
+                pop.name,
+                description=pop.description if pop.description else None,
+                start_time=math.inf,
+                epochs=[initial_epoch],
+            )
+
+        deme_map = dict()
+        for pop in resolved.populations:
+            deme_map[pop.name] = b.data["demes"][pop.id]
+            deme_map[pop.id] = b.data["demes"][pop.id]
+
+        # Copied from demes/ms.py
+        def epoch_resolve(deme, time):
+            """
+            Return the oldest epoch if it has end_time == time. If not, create a
+            new oldest epoch with end_time=time. Also resolve sizes by dealing
+            with the growth_rate attribute (if required).
+            """
+            epoch = deme["epochs"][0]
+            start_time = deme["start_time"]
+            end_time = epoch["end_time"]
+            if not (start_time > time >= end_time):
+                raise ValueError(
+                    f"time {time} outside {deme['name']}'s existence interval "
+                    f"(start_time={start_time}, end_time={end_time}]"
+                )
+
+            if time > end_time:
+                new_epoch = copy.deepcopy(epoch)
+                # find size at given time
+                growth_rate = epoch.pop("growth_rate", 0)
+                dt = time - epoch["end_time"]
+                size_at_t = epoch["end_size"] * math.exp(-growth_rate * dt)
+                epoch["start_size"] = size_at_t
+                new_epoch["end_size"] = size_at_t
+                new_epoch["end_time"] = time
+                deme["epochs"].insert(0, new_epoch)
+                epoch = new_epoch
+
+            return epoch
+
+        # List of migration matrices in time-descending order (oldest to most recent).
+        mm_list = [resolved.migration_matrix.T]
+        # End times for each migration matrix.
+        mm_end_times = [0.0]
+
+        # Copied from demes/ms.py
+        def migration_matrix_at(time):
+            """
+            Return the oldest migration matrix if it has end_time == time. If not,
+            create a new oldest migration matrix with end_time = time.
+            """
+            migration_matrix = mm_list[0]
+            if time > mm_end_times[0]:
+                # We need a new migration matrix.
+                migration_matrix = copy.deepcopy(migration_matrix)
+                mm_list.insert(0, migration_matrix)
+                mm_end_times.insert(0, time)
+            return migration_matrix
+
+        for time, events_group in itertools.groupby(
```

Comment:
There's a lot of logic in this loop... I've found it a lot easier in things like ``assert_equivalent`` to let the debugger do some of the heavy lifting. It saves a lot of reasoning about the effects of events, because you can just detect the change in state of the model. I think this makes things like changing migration rates a lot easier, and definitely detecting the start/end times of Demes (you can just look at the state of the Population in each epoch).

Any thoughts?

Reply:
Ah, yes. The migration matrices from the debugger should be a drop in replacement. I'll take a look.